src/core_plugins/kibana/public/dashboard/dashboard_state_manager.js

```diff
@@ -167,10 +167,25 @@ export class DashboardStateManager {
     // AppState, which then dispatches the change here, which will end up triggering setState warnings.
     if (!this._areStoreAndAppStatePanelsEqual()) {
       this.triggerEmbeddableConfigUpdateListeners();
-      store.dispatch(setPanels(this.getPanels()));
+
+      const panelsMap = {};
+      this.getPanels().forEach(panel => {
+        panelsMap[panel.panelIndex] = { ...panel };
+      });
+
+      store.dispatch(setPanels(panelsMap));
     }
 
     const state = store.getState();
+
+    if (getTitle(state) !== this.getTitle()) {
+      store.dispatch(updateTitle(this.getTitle()));
```

Comment:
It seems we have 2 sources of truth. Is this because we're synchronizing between a React/redux app and an Angular one?

Reply:
Yes, it sucks!  appState is in the angular world, but it really just represents the url.  I want to eventually remove appState and introduce react router.  But I've also read that one shouldn't consider the url as a source of truth, rather just as another ui element displaying the source of truth of your redux state.  But we first read from the url, so it's like url -> appState -> redux.  Then the url can be modified directly, so we need to push those changes into redux, or the opposite way (ui control, from redux into appState/the url).

Even if I switch appState with react router, I think I'll essentially be in the same boat. If you have ideas on improving this I'd love to have a conversation about it!

Reply:
Hm. Yeah. That's tricky. I've always treated the URL as the source of truth. I know there are some frameworks (in Clojure, at least) that do this as well, so it's not objectively wrong to do that! I'll have to think about it!

Reply:
I should clarify because I just realized appState does not quite mean url state.  It's the `_a` parameter in the url.  It contains most of the state needed for an application, but there is also `_g` which is "global state" and that contains the time filter data.